apps/sync/js/fxsync-webcrypto/fxsyncwebcrypto.js

```diff
@@ -0,0 +1,279 @@
+     cryptoKeysCiphertextByteArray) {
+    return crypto.subtle.decrypt({ name: 'AES-CBC', iv: cryptoKeysIVByteArray },
+        this.mainSyncKey.aes,
+        cryptoKeysCiphertextByteArray).then(function(keyBundleAB) {
+      var cryptoKeysJSON = String.fromCharCode.apply(null,
+          new Uint8Array(keyBundleAB));
+      try {
+        this.bulkKeyBundle = JSON.parse(cryptoKeysJSON);
+        return importKeyBundle(
+            StringConversion.base64StringToByteArray(
+                this.bulkKeyBundle.default[0]),
+            StringConversion.base64StringToByteArray(
+                this.bulkKeyBundle.default[1])
+        ).then(function(keyBundle) {
+          this.bulkKeyBundle.defaultAsKeyBundle = keyBundle;
+        }.bind(this));
+      } catch(e) {
+        return Promise.reject('Deciphered crypto keys, but not JSON');
+      }
+    }.bind(this), function(err) {
+      return Promise.reject(
+          'Could not decrypt crypto keys using AES part of stretched kB key');
+    });
+  };
+
+  function importFromStrings(obj) {
+    var ret = {};
+    try {
+      ret.ciphertext = StringConversion.base64StringToByteArray(
+          obj.ciphertext);
+    } catch (e) {
+      throw new Error(
+          'Could not parse ciphertext as a base64 string');
+    }
+
+    // Intentionally using StringConversion.rawStringToByteArray
+    // instead of StringConversion.base64StringToByteArray on the ciphertext
+    // here - See https://github.com/mozilla/firefox-ios/blob/ \
+    // 1cce59c8eac282e151568f1204ffbbcc27349eff/Sync/KeyBundle.swift#L178
+    ret.hmacSignedText = StringConversion.rawStringToByteArray(
+          obj.ciphertext);
+
+    try {
+      ret.IV =
+          StringConversion.base64StringToByteArray(obj.IV);
+    } catch (e) {
+      throw new Error('Could not parse IV as a base64 string');
+    }
+    try {
+      ret.hmacSignature =
+          StringConversion.hexStringToByteArray(obj.hmac);
+    } catch (e) {
+      throw new Error('Could not parse hmac as a hex string');
+    }
+    return ret;
+  }
+
+  /*
+   * setKeys - import kB and crypto/keys
+   *
+   * @param {String} kB Hex string with kB from FxA onepw protocol
+   * @param {Object} cryptoKeysStrings Object with:
+   *     - ciphertext {String} A Base64 String containing an AES-CBC ciphertext
+   *     - IV {String} A Base64 String containing the AES-CBC Initialization
+   *         Vector
+   *     - hmac {String} A Hex String containing the HMAC-SHA256 signature
+   * @returns {Promise} A promise that will resolve after import of kB and
+   *     decryption of cryptoKeys.
+   */
+  FxSyncWebCrypto.prototype.setKeys = function(kB, cryptoKeysStrings) {
+    var kBByteArray, cryptoKeys;
+
+    // Input checking
+    try {
+      kBByteArray = StringConversion.hexStringToByteArray(kB);
+    } catch (e) {
+      return Promise.reject('Could not parse kB as a hex string');
+    }
+
+    cryptoKeys = importFromStrings(cryptoKeysStrings);
```

Comment:
Actually, that's nonsense, because the exception would be thrown on the first line of that snippet, before chai-as-expected even gets near it. That test passes because it was testing something else. :) Errors thrown by `importFromStrings` are  a missing test case. Will add it.